x-pack/plugins/ml/server/models/data_recognizer/data_recognizer.ts

```diff
@@ -306,81 +333,96 @@ export class DataRecognizer {
     const datafeeds: ModuleDatafeed[] = [];
     const kibana: KibanaObjects = {};
     // load all of the job configs
-    await Promise.all(
-      manifestJSON.jobs.map(async (job) => {
-        try {
-          const jobConfig = await this.readFile(
-            `${this._modulesDir}/${dirName}/${ML_DIR}/${job.file}`
-          );
-          // use the file name for the id
-          jobs.push({
-            id: `${prefix}${job.id}`,
-            config: JSON.parse(jobConfig),
-          });
-        } catch (error) {
-          mlLog.warn(
-            `Data recognizer error loading config for job ${job.id} for module ${id}. ${error}`
-          );
-        }
-      })
-    );
-
-    // load all of the datafeed configs
-    await Promise.all(
-      manifestJSON.datafeeds.map(async (datafeed) => {
-        try {
-          const datafeedConfig = await this.readFile(
-            `${this._modulesDir}/${dirName}/${ML_DIR}/${datafeed.file}`
-          );
-          const config = JSON.parse(datafeedConfig);
-          // use the job id from the manifestFile
-          config.job_id = `${prefix}${datafeed.job_id}`;
-
-          datafeeds.push({
-            id: prefixDatafeedId(datafeed.id, prefix),
-            config,
-          });
-        } catch (error) {
-          mlLog.warn(
-            `Data recognizer error loading config for datafeed ${datafeed.id} for module ${id}. ${error}`
-          );
-        }
-      })
-    );
+    if (isModule(module)) {
+      const tempJobs = module.jobs.map((j) => ({ id: `${prefix}${j.id}`, config: j.config }));
+      jobs.push(...((tempJobs as unknown) as ModuleJob[]));
+      const tempDatafeeds = module.datafeeds.map((d) => ({
+        id: prefixDatafeedId(d.id, prefix),
+        config: {
+          ...d.config,
+          job_id: `${prefix}${d.job_id}`,
+        },
+      }));
+      datafeeds.push(...((tempDatafeeds as unknown) as ModuleDatafeed[]));
+    } else if (isFileBasedModule(module)) {
+      await Promise.all(
+        module.jobs.map(async (job) => {
+          try {
+            const jobConfig = await this._readFile(
+              `${this._modulesDir}/${dirName}/${ML_DIR}/${job.file}`
+            );
+            // use the file name for the id
+            jobs.push({
+              id: `${prefix}${job.id}`,
+              config: JSON.parse(jobConfig),
+            });
+          } catch (error) {
+            mlLog.warn(
+              `Data recognizer error loading config for job ${job.id} for module ${id}. ${error}`
+            );
+          }
+        })
+      );
```

Comment:
I think that the `.map` method is misused. Semantically it means we create a new collection based on the other one, but here the callback contains a side-effect. I would use `for...of` in this case. If you want to utilize `Promise.all`, it might be `.reduce` instead. It's relevant for `module.datafeeds` and `module!.kibana[key]` as well

Reply:
i've updated the code for `jobs` and `datafeeds`, however I'm reluctant to change the `kibana` objects loop due to how complicated it is.
I agree the side effects in the `map` are not ideal, however the code is very old and works correctly. Plus in the next stage of this work these loops will be removed completely.

Reply:
https://github.com/elastic/kibana/pull/92855/commits/961396d595104dfe9d633151737a81fde9764366